staging/src/k8s.io/apimachinery/pkg/util/waitgroup/ratelimited_waitgroup_test.go

```diff
@@ -0,0 +1,320 @@
+	// and invoke 'Done' to finish using this waitgroup.
+	close(blockedCh)
+	n2DoneWG.Wait()
+
+	if invoked := limiter.invoked(); invoked != n2 {
+		t.Errorf("expected rate limiter to be called %d times, but got: %d", n2, invoked)
+	}
+
+	waitResult := <-waitDoneCh
+	if count := target.Count(); count != 0 {
+		t.Errorf("expected count to be zero, but got: %d", count)
+	}
+	if waitResult.before != n2 {
+		t.Errorf("expected count before Wait to be: %d, but got: %d", n2, waitResult.before)
+	}
+	if waitResult.after != 0 {
+		t.Errorf("expected count after Wait to be zero, but got: %d", waitResult.after)
+	}
+	if cancelInvoked != 1 {
+		t.Errorf("expected context cancel to be invoked once, but got: %d", cancelInvoked)
+	}
+}
+
+func TestRateLimitedSafeWaitGroupWithHardTimeout(t *testing.T) {
+	target := &rateLimitedSafeWaitGroupWrapper{
+		RateLimitedSafeWaitGroup: &RateLimitedSafeWaitGroup{},
+	}
+	n := 10
+	wg := sync.WaitGroup{}
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			target.Add(1)
+		}()
+	}
+
+	wg.Wait()
+	if count := target.Count(); count != n {
+		t.Errorf("expected count to be: %d, but got: %d", n, count)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	activeAt, activeNow, err := target.Wait(func(count int) (RateLimiter, context.Context, context.CancelFunc) {
+		return nil, ctx, cancel
+	})
+	if activeAt != n {
+		t.Errorf("expected active at Wait to be: %d, but got: %d", n, activeAt)
+	}
+	if activeNow != n {
+		t.Errorf("expected active after Wait to be: %d, but got: %d", n, activeNow)
+	}
+	if err != context.Canceled {
+		t.Errorf("expected error: %v, but got: %v", context.Canceled, err)
+	}
+}
+
+func TestRateLimitedSafeWaitGroupWithBurstOfOne(t *testing.T) {
+	target := &rateLimitedSafeWaitGroupWrapper{
+		RateLimitedSafeWaitGroup: &RateLimitedSafeWaitGroup{},
+	}
+	n := 200
+	grace := 5 * time.Second
+	wg := sync.WaitGroup{}
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			target.Add(1)
+		}()
+	}
+	wg.Wait()
+
+	waitingCh := make(chan struct{})
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+
+			<-waitingCh
+			target.Done()
+		}()
+	}
+	defer wg.Wait()
+
+	now := time.Now()
+	t.Logf("Wait starting, N=%d, grace: %s, at: %s", n, grace, now)
+	activeAt, activeNow, err := target.Wait(func(count int) (RateLimiter, context.Context, context.CancelFunc) {
+		defer close(waitingCh)
+		// no deadline in context, Wait will wait forever, we want to measure
+		// how long it takes for the requests to drain.
+		return rate.NewLimiter(rate.Limit(n/int(grace.Seconds())), 1), context.Background(), func() {}
+	})
+	took := time.Since(now)
+	t.Logf("Wait finished, count(before): %d, count(after): %d, took: %s, err: %v", activeAt, activeNow, took, err)
+
+	// in CPU starved environment, the go routines may not finish in time
+	if took > 2*grace {
+		t.Errorf("expected Wait to take: %s, but it took: %s", grace, took)
```

Comment:
@wojtek-t  This is a new test I added, I am allowing the test to take double the time allowed, in case the test runner is under stress, if this is not enough, I can remove this.

Reply:
I think this should be fine.